lib/web/apiserver_test.go

```diff
@@ -2008,6 +2009,142 @@ func TestCreatePrivilegeToken(t *testing.T) {
 	require.NotEmpty(t, privilegeToken)
 }
 
+func TestCreateTOTPSecrets(t *testing.T) {
+	t.Parallel()
+	env := newWebPack(t, 1)
+	proxy := env.proxies[0]
+
+	ctx := context.Background()
+	username := "foo@example.com"
+	pack := proxy.authPack(t, username)
+
+	// Acquire a user token.
+	token, err := types.NewUserToken("some-token-id")
+	require.NoError(t, err)
+	token.SetUser(username)
+	token.SetExpiry(env.clock.Now().Add(5 * time.Minute))
+	_, err = env.server.Auth().Identity.CreateUserToken(ctx, token)
+	require.NoError(t, err)
+
+	endpoint := pack.clt.Endpoint("webapi", "mfa", "token", token.GetName(), "totpsecrets")
+	re, err := pack.clt.PostJSON(ctx, endpoint, url.Values{})
+	require.NoError(t, err)
+
+	var qrcode []byte
+	err = json.Unmarshal(re.Bytes(), &qrcode)
+	require.NoError(t, err)
+	require.NotEmpty(t, qrcode)
+}
+
+func TestAddMFADevice(t *testing.T) {
+	t.Parallel()
+	ctx := context.Background()
+	env := newWebPack(t, 1)
+	proxy := env.proxies[0]
+	pack := proxy.authPack(t, "foo@example.com")
+
+	// Enable second factor on.
+	ap, err := types.NewAuthPreference(types.AuthPreferenceSpecV2{
+		Type:         constants.Local,
+		SecondFactor: constants.SecondFactorOptional,
+		U2F: &types.U2F{
+			AppID:  "https://" + env.server.ClusterName(),
+			Facets: []string{"https://" + env.server.ClusterName()},
+		},
+	})
+	require.NoError(t, err)
+	err = env.server.Auth().SetAuthPreference(ctx, ap)
+	require.NoError(t, err)
+
+	// Get a totp code to re-auth.
+	totpCode, err := totp.GenerateCode(pack.otpSecret, env.clock.Now().Add(30*time.Second))
+	require.NoError(t, err)
+
+	// Obtain a privilege token.
+	endpoint := pack.clt.Endpoint("webapi", "users", "privilege", "token")
+	re, err := pack.clt.PostJSON(ctx, endpoint, &privilegeTokenRequest{
+		SecondFactorToken: totpCode,
+	})
+	require.NoError(t, err)
+	var privilegeToken string
+	require.NoError(t, json.Unmarshal(re.Bytes(), &privilegeToken))
+
+	tests := []struct {
+		name        string
+		deviceName  string
+		getTOTPCode func() string
+		getU2FResp  func() *fidou2f.RegisterResponse
```

Comment:
Please use lib/auth/u2f.RegisterChallengeResponse (and other lib/auth/u2f aliases where applicable). We should not be importing tstranex/u2f outside of specialized packages (that really make use of it).